codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XMLProtocolUtils.java

```diff
@@ -267,9 +269,30 @@ public static void writeXmlErrorMessageCodeDeserializer(ProtocolGenerator.Genera
         boolean isNoErrorWrapping = context.getService().getTrait(RestXmlTrait.class).map(
                 RestXmlTrait::isNoErrorWrapping).orElse(false);
 
-        writer.addUseImports(SmithyGoDependency.SMITHY_XML);
-        writer.write("errorCode, err := smithyxml.GetResponseErrorCode(errorBody, $L)", isNoErrorWrapping);
-        writer.write("if err != nil { return err }");
+        ServiceShape service = context.getService();
+
+        writer.write("m := middleware.Metadata{}");
+        // TODO: should this be a Runtime client plugin?
+        if (service.getId().getName() == "S3") {
+            writer.addUseImports(AwsCustomGoDependency.S3_CUSTOMIZATION);
+            writer.write("errorComponents, err := s3cust.GetErrorResponseComponents(errorBody)");
+            writer.write("if err != nil { return m, err }");
+            writer.write("s3cust.SetHostIDMetadata(&m, errorComponents.HostID)");
+        } else {
+            writer.addUseImports(AwsGoDependency.AWS_XML);
+            writer.write("errorComponents, err := awsxml.GetErrorResponseComponents(errorBody, $L)", isNoErrorWrapping);
+            writer.write("if err != nil { return m, err }");
+        }
+
+        writer.addUseImports(AwsGoDependency.AWS_HTTP_TRANSPORT);
+        writer.write("awshttp.SetRequestIDMetadata(&m, errorComponents.RequestID)");
+        writer.insertTrailingNewline();
+
+        // assign middleware metadata to metadata var
+        writer.write("metadata = m");
```

Comment:
Same comment as non-s3 case, why duplicate/replace metadata?

How does this work with the operation middleware for `deserMetadata` variable created by smithy-go?

Reply:
The metadata here isn't the middleware.Metadata that is being passed around in operation stack. 

The naming should be changed to prevent the confusion. The metadata here is the return argument (an empty interface) for top level error deserializer function.

Reply:
metadata is actually the `deserMetadata`. It is used as a return argument here